tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/LocaleCompletion.java

```diff
@@ -83,39 +95,68 @@ public class LocaleCompletion {
+        @PathParam("locale") @Schema(required = true, description = "Locale ID", example = "aa") String localeId) throws ExecutionException {
+        if (SurveyMain.isBusted() || !SurveyMain.wasInitCalled() || !SurveyMain.triedToStartUp()) {
+            return STError.surveyNotQuiteReady();
+        }
         CLDRLocale cldrLocale = CLDRLocale.getInstance(localeId);
-        return getLocaleCompletion(cldrLocale);
+        return Response.ok(getLocaleCompletion(cldrLocale)).build();
+    }
+
+    static LocaleCompletionResponse getLocaleCompletion(CLDRLocale cldrLocale) throws ExecutionException {
+        return LocaleCompletionHelper.INSTANCE.cache.get(cldrLocale);
+        // return handleGetLocaleCompletion(cldrLocale);
     }
 
     /**
-     * Fetch locale completion for the specified locale.
-     * Note: Static and package-private for testability.
+     * This function computes the actual locale completion given a Locale
      * @param cldrLocale
-     * @return Response
+     * @return
      */
-    static Response getLocaleCompletion(CLDRLocale cldrLocale) {
-        if (SurveyMain.isBusted() || !SurveyMain.wasInitCalled() || !SurveyMain.triedToStartUp()) {
-            return STError.surveyNotQuiteReady();
-        }
-
+    static LocaleCompletionResponse handleGetLocaleCompletion(CLDRLocale cldrLocale) {
         final STFactory stFactory = CookieSession.sm.getSTFactory();
-        final LocaleCompletionResponse lcr = getLocaleCompletion(cldrLocale, stFactory);
-        return Response.ok(lcr).build();
+        final LocaleCompletionResponse lcr = handleGetLocaleCompletion(cldrLocale, stFactory);
+        return lcr;
     }
 
-    static final class LocaleCompletionHelper {
+    static final class LocaleCompletionHelper implements Listener {
         PathHeader.Factory phf = null;
-
+        LoadingCache<CLDRLocale, LocaleCompletionResponse> cache;
         LocaleCompletionHelper() {
             phf = PathHeader.getFactory(CLDRConfig.getInstance().getEnglish());
+            cache =  CacheBuilder.newBuilder().maximumSize(500)
+            // We would want the expiry time greater if this were used by users, not just TC.
+            .expireAfterWrite(15, TimeUnit.MINUTES)
+            // .concurrencyLevel()
+            .build(new CacheLoader<CLDRLocale, LocaleCompletionResponse>() {
+                @Override
+                public LocaleCompletionResponse load(CLDRLocale key) throws Exception {
+                    return handleGetLocaleCompletion(key);
+                }
+            });
         }
-
         static LocaleCompletionHelper INSTANCE = new LocaleCompletionHelper();
+
+        @Override
+        public void valueChanged(String xpath, XMLSource source) {
+            // invalidate the named entry
+            // this is registered whenever a calculation happens
+            System.err.println("LCH Invalidating " + source.getLocaleID());
```

Comment:
System.err.println -- probably should be only when debugging? and "info" not "err"?